src/metabase/util/analytics.clj

```diff
@@ -0,0 +1,119 @@
+(ns metabase.util.analytics
+  "Functions for sending Snowplow analytics events"
+  (:require [clojure.tools.logging :as log]
+            [metabase.config :as config]
+            [metabase.models.setting :refer [defsetting]]
+            [metabase.public-settings :as public-settings]
+            [metabase.util.i18n :as i18n :refer [deferred-tru trs]])
+  (:import [com.snowplowanalytics.snowplow.tracker Subject$SubjectBuilder Tracker Tracker$TrackerBuilder]
+           [com.snowplowanalytics.snowplow.tracker.emitter BatchEmitter BatchEmitter$Builder Emitter]
+           [com.snowplowanalytics.snowplow.tracker.events Unstructured Unstructured$Builder]
+           [com.snowplowanalytics.snowplow.tracker.http ApacheHttpClientAdapter ApacheHttpClientAdapter$Builder]
+           com.snowplowanalytics.snowplow.tracker.payload.SelfDescribingJson
+           org.apache.http.impl.client.HttpClients
+           org.apache.http.impl.conn.PoolingHttpClientConnectionManager))
+
+(defsetting snowplow-url
+  (deferred-tru "The URL of the Snowplow collector to send analytics events to")
+  :default (if config/is-prod?
+             "https://sp.metabase.com"
+             ;; See the iglu-schema-registry repo for instructions on how to run Snowplow Micro locally for development
+             "http://localhost:9095")
+  :visibility :public)
+
+(def ^:private ^Emitter emitter
```

Comment:
adding the `^Emitter` tag is incorrect here because this is actually going to be a `clojure.lang.Delay` rather than an `Emitter`. In this case you might want to do something like

```clj
(def ^:private ^{:arglists `(^Emitter [])} emitter
  (let [emitter* (delay ...)]
    (fn [] @emitter*)))
```

Then you can call `(emitter)` which will give you the instance. Note that you want to do the `:arglists` with a backtick so it qualifies/resolves `Emitter` which will give it the `:tag` `com.snowplowanalytics.snowplow.tracker.emitter.Emitter`. e.g. we should get something like this:

```clj
(-> #'emitter meta :arglists first meta :tag)
;; -> com.snowplowanalytics.snowplow.tracker.emitter.Emitter ; symbol, not the class
```

Reply:
The `:argslist` metadata causes an exception to be thrown in the Eastwood linter:
```
Exception thrown during phase :analyze+eval of linting namespace metabase.analytics.snowplow
UnsupportedOperationException count not supported on this type: Symbol
```

I don't understand the root cause, but it seems that removing the type hints here (and on `tracker`) doesn't cause any reflection warnings. So I've removed them for now.